src/Database/Driver/Sqlserver.php

```diff
@@ -187,12 +187,15 @@ public function enabled(): bool
      * Prepares a sql statement to be executed
      *
      * @param string|\Cake\Database\Query $query The query to prepare.
+     * @param array $options The driver options to set on the statement.
      * @return \Cake\Database\StatementInterface
      */
-    public function prepare($query): StatementInterface
+    public function prepare($query, array $options = []): StatementInterface
     {
         $this->connect();
-        $options = [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL];
+        if (!isset($options[PDO::ATTR_CURSOR]) {
+            $options = [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL] + $options;
+        }
```

Comment:
```suggestion
    $options = $options + [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL];
```

Reply:
@othercorey if we do that the `PDO::ATTR_CURSOR` will be overridden, but it will be appended, being the last index of the array. Unfortunately, this would throw the exception:

>SQLSTATE[IMSSP]: The PDO::SQLSRV_ATTR_CURSOR_SCROLL_TYPE attribute may only be set when PDO::ATTR_CURSOR is set to PDO::CURSOR_SCROLL in the $driver_options array of PDO::prepare.

I had it happening to me, seems the `PDO::ATTR_CURSOR` must be set before the other options... hence why I used that check...

Reply:
My mistake, I forgot it was an integer constant.

Reply:
Do you always want ATTR_CURSOR set to something? If don't need to add the defaults if $options is not empty.

Reply:
Actually not... there can be many cases it won't even be required, for instance when you want to fetch all results without iterating.

However I left it there as it was set by default by the framework, so perhaps it's required for the usage of cursor methods later on some other Cake Sqlserver classes.

Reply:
Or you mean something like..

```php
if (empty($options)) {
    $options = [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL]
}
```

or just set it on the parameter as default?

```php
public function prepare($query, array $options = [PDO::ATTR_CURSOR => PDO::CURSOR_SCROLL]): StatementInterface
```

Reply:
Yes, like the one with only default set. I don't know the history of CURSOR_SCROLL in CakePHP though.

Reply:
Yeah, to be honest I have no idea either... that's why I left it that way. 😅

I traced it back to 2014 at this commit https://github.com/cakephp/cakephp/commit/3bf580dbe72a278ba2c0276504ac69e5877087b1#diff-6d58c9b7093fe3eba95388ad543a0da1R104 seems it's a default since then.

Reply:
Can you run a test with that attribute removed? I don't know where CakePHP need to scroll backwards in the cursor.

Reply:
@othercorey I tried without any parameters for cursor set and it throws the following error:
> array size cannot be less than zero
> InvalidArgumentException 

**Trace:**
https://github.com/cakephp/cakephp/blob/47912247fa64dc35066c8fe74d24bee38480d4b6/src/ORM/ResultSet.php#L177
https://github.com/cakephp/cakephp/blob/47912247fa64dc35066c8fe74d24bee38480d4b6/src/ORM/Query.php#L1114
https://github.com/cakephp/cakephp/blob/47912247fa64dc35066c8fe74d24bee38480d4b6/src/Datasource/QueryTrait.php#L292
https://github.com/cakephp/cakephp/blob/47912247fa64dc35066c8fe74d24bee38480d4b6/src/ORM/Query.php#L1059
https://github.com/cakephp/cakephp/blob/47912247fa64dc35066c8fe74d24bee38480d4b6/src/Datasource/QueryTrait.php#L141

Although, following the trace seems I'd need to readapt some code to avoid this.

Wherever I use ORM to find Database results I'd need to call ORM's `disableBufferedResults` method:

```php
$users = $this->getTableLocator()->get('Users')->find('all')->disableBufferedResults();
```

Once I did that it worked flawlessly. I will check how the database receives the EXEC queries, however.

Reply:
@othercorey  indeed, I've just tested that and it was much more performatic as well, and the query was received by the server as it was sent.

<img width="461" alt="Screen Shot 2020-07-21 at 00 33 36" src="https://user-images.githubusercontent.com/1733348/88011099-dbfb6600-caec-11ea-89ff-42b0bbdc091d.png">

Ran three procedures with the settings:
**red** is using `CURSOR_SCROLL`
**green** is using `CURSOR_FWDONLY`

The `CURSOR_SCROLL` enforces the usage of `sp_cursorprepexec` which runs a lot of cursor operations to create a handle, fetch result, and return.


According to [Microsoft documentation](https://docs.microsoft.com/en/sql/connect/php/pdo-prepare?view=sql-server-ver15#remarks) the default is `CURSOR_FWDONLY`.
Even since when I use the `CURSOR_SCROLL` and `SQLSRV_CURSOR_BUFFERED` seems equally performatic as `CURSOR_FWDONLY`, I will keep only `CURSOR_FWDONLY` on your app.

Reply:
@othercorey I think it's safe to say rather set Cake's default also to `CURSOR_FWDONLY`? 😅

Reply:
Probably. Like I said earlier - I don't know where cakephp needs to scroll backwards so FWDONLY should work.